arrows/klv/klv_read_write.txx

```diff
@@ -0,0 +1,378 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+/// \file
+/// Implementation of the templated basic KLV read/write functions.
+
+#include "klv_read_write.h"
+
+#include <vital/exceptions.h>
+
+#include <limits>
+#include <numeric>
+#include <string>
+
+#include <cmath>
+
+namespace kwiver {
+  namespace arrows {
+    namespace klv {
+// ---------------------------------------------------------------------------
+// Macros for verifying type attributes
+// Message-less static_assert not available in C++11
+
+#define KLV_ASSERT_INT( T ) \
+  static_assert( std::is_integral < T > ::value, "must be an integer type" )
+
+#define KLV_ASSERT_UINT( T )                                                \
+  KLV_ASSERT_INT( T );                                                      \
+  static_assert( std::is_unsigned < T > ::value, "must be an unsigned type" )
+
+#define KLV_ASSERT_SINT( T )                                           \
+  KLV_ASSERT_INT( T );                                                 \
+  static_assert( std::is_signed < T > ::value, "must be a signed type" )
+
+#define KLV_ASSERT_UINT8_ITERATOR( ITER )                              \
+  static_assert(                                                       \
+    std::is_same < typename std::decay < decltype( *ITER ) > ::type,      \
+    uint8_t > ::value, "iterator must point to uint8_t" )
+
+// ---------------------------------------------------------------------------
+// Returns number of bits required to store the given signed or unsigned int.
+      template < class T >
+      KWIVER_ALGO_KLV_EXPORT
+      size_t
+      _int_bit_requirement( T value );
```

Comment:
What is the purpose of forward declarations? Why do these start with `_`? If they're meant to be internal, putting them in an anonymous namespace (and also a `.cxx`) might be better?

Reply:
...Huh. I suppose forward declarations aren't necessary for use within template functions. My bad.

I don't think putting them in a `.cxx` would work because they're templated and used within templates. Doesn't the compiler need the full template definition in the header? You are correct that the `_` was meant to mark them as internal, since (correct me if I'm wrong here) an anonymous namespace wouldn't make sense in a header. An alternative would be explicitly instantiating the helper templates for all foreseen types, then moving the definitions into a `.cxx`, but that seems less elegant to me.

Reply:
Turns out that removing forward declarations does cause a compile error. Even with templates, functions need to be declared before use.

Reply:
Let me rephrase that, then; is there a reason your definitions aren't in order of dependence? 🙂